File: apps/web/src/api/researchProducts.ts
/**
 * 研究产物 API 客户端（阶段 3 新增）
 *
 * 端点列表：
 *   GET    /research/workspaces/{id}/runs/{runId}/candidates                    — 列出候选产物
 *   GET    /research/workspaces/{id}/runs/{runId}/candidates/{candidateId}      — 候选详情
 *   POST   /research/workspaces/{id}/derived-datasets                           — 创建数据集
 *   GET    /research/workspaces/{id}/derived-datasets                           — 列出数据集
 *   GET    /research/workspaces/{id}/derived-datasets/{datasetId}              — 数据集详情
 *   PATCH  /research/workspaces/{id}/derived-datasets/{datasetId}              — 编辑元数据
 *   GET    /research/workspaces/{id}/derived-datasets/{datasetId}/versions     — 版本历史
 *   GET    /research/workspaces/{id}/derived-datasets/{datasetId}/versions/{vn} — 版本详情
 *   POST   /research/workspaces/{id}/views                                      — 创建视图
 *   GET    /research/workspaces/{id}/views                                      — 列出视图
 *   GET    /research/workspaces/{id}/views/{viewId}                            — 视图详情
 *   PATCH  /research/workspaces/{id}/views/{viewId}                            — 编辑元数据
 *   GET    /research/workspaces/{id}/views/{viewId}/versions                   — 版本历史
 *   GET    /research/workspaces/{id}/views/{viewId}/versions/{vn}              — 版本详情
 *   GET    /research/workspaces/{id}/views/{viewId}/versions/{vn}/image        — 下载图片
 *   GET    /research/workspaces/{id}/insights                                   — 列出 Insight
 *   GET    /research/workspaces/{id}/insights/{insightId}                     — Insight 详情
 *   PATCH  /research/workspaces/{id}/insights/{insightId}                     — 编辑元数据
 *   GET    /research/workspaces/{id}/insights/{insightId}/versions           — 版本历史
 *   GET    /research/workspaces/{id}/runs/{runId}/insight-candidates          — 列出候选
 *   GET    /research/workspaces/{id}/runs/{runId}/insight-candidates/{cid}   — 候选详情
 *   POST   /research/workspaces/{id}/runs/{runId}/insight-candidates/{cid}/accept — 接受
 *   POST   /research/workspaces/{id}/runs/{runId}/insight-candidates/{cid}/modify — 修改
 *   POST   /research/workspaces/{id}/runs/{runId}/insight-candidates/{cid}/reject — 拒绝
 *   GET    /research/workspaces/{id}/products                                  — 产物列表
 *   GET    /research/catalog/search                                            — 搜索衍生数据
 *
 * 风格参考 apps/web/src/api/research.ts：纯 async 函数 + http 实例。
 */
import { http } from './client';

// ============================================================
// 类型定义
// ============================================================

export type CandidateProduct = {
  candidate_type: string;
  source_artifact_id: string | null;
  candidate_id: string;
  source_run_id: string;
  source_step_id: string | null;
  step_name: string;
  step_status: string;
  preview_data: Record<string, unknown>;
  status: string;
  error_reason: string;
};

export type CandidateListResponse = {
  items: CandidateProduct[];
};

export type CandidateDetailResponse = {
  candidate_type: string;
  candidate_id: string;
  source_run_id: string;
  source_step_id: string | null;
  preview_data: Record<string, unknown>;
};

export type DerivedDataset = {
  dataset_id: string;
  name: string;
  status: string;
  current_version: number;
  workspace_id: string;
};

export type DatasetListResponse = {
  items: DerivedDataset[];
};

export type DatasetDetail = {
  dataset_id: string;
  workspace_id: string;
  name: string;
  summary: string | null;
  tags: string[];
  status: string;
  current_version: number;
  source_run_id: string;
  source_snapshot_id: string | null;
  current_version_data: Record<string, unknown> | null;
};

export type DatasetVersion = {
  version_id: string;
  dataset_id: string;
  version_number: number;
  content_hash: string;
  created_at: string;
};

export type DatasetVersionListResponse = {
  items: DatasetVersion[];
};

export type DatasetVersionDetail = {
  version_id: string;
  dataset_id: string;
  version_number: number;
  metadata_content: Record<string, unknown>;
  points_content: Array<Record<string, unknown>>;
  series_content: Array<Record<string, unknown>>;
  field_manifest: Array<Record<string, unknown>>;
  content_hash: string;
  source_run_id: string;
  source_step_id: string | null;
  source_artifact_id: string | null;
  created_at: string;
};

export type View = {
  view_id: string;
  name: string;
  status: string;
  current_version: number;
  caption: string | null;
  display_order: number;
};

export type ViewListResponse = {
  items: View[];
};

export type ViewDetail = {
  view_id: string;
  workspace_id: string;
  name: string;
  caption: string | null;
  display_order: number;
  status: string;
  current_version: number;
  source_run_id: string;
  current_version_info: Record<string, unknown> | null;
};

export type ViewVersion = {
  version_id: string;
  view_id: string;
  version_number: number;
  image_storage_path: string;
  image_format: string;
  created_at: string;
};

export type ViewVersionListResponse = {
  items: ViewVersion[];
};

export type ViewVersionDetail = {
  version_id: string;
  view_id: string;
  version_number: number;
  image_storage_path: string;
  image_format: string;
  image_width: number | null;
  image_height: number | null;
  image_content_hash: string;
  chart_code_artifact_id: string | null;
  image_digest: string | null;
  source_run_id: string;
  source_step_id: string | null;
  source_artifact_id: string | null;
  bound_dataset_version_id: string | null;
  chart_description: string | null;
  created_at: string;
};

export type Insight = {
  insight_id: string;
  name: string;
  status: string;
  current_version: number;
};

export type InsightListResponse = {
  items: Insight[];
};

export type InsightDetail = {
  insight_id: string;
  workspace_id: string;
  name: string;
  status: string;
  current_version: number;
  source_run_id: string | null;
  current_version_data: Record<string, unknown> | null;
};

export type InsightVersion = {
  version_id: string;
  insight_id: string;
  version_number: number;
  is_modified: boolean;
  created_at: string;
};

export type InsightVersionListResponse = {
  items: InsightVersion[];
};

export type InsightCandidate = {
  candidate_id: string;
  run_id: string;
  step_id: string | null;
  status: string;
  conclusion: string;
  evidence_source_label: string;
  created_at: string;
};

export type InsightCandidateListResponse = {
  items: InsightCandidate[];
};

export type ProductSummary = {
  product_type: string;
  product_id: string;
  name: string;
  status: string;
  current_version: number;
};

export type ProductListResponse = {
  items: ProductSummary[];
};

export type CatalogSearchResult = {
  id: string;
  name: string;
  current_version: number;
  workspace_id: string;
  owner_user_id: string;
  summary: string;
  tags: string[];
};

export type CatalogSearchResponse = {
  items: CatalogSearchResult[];
};

// ============================================================
// API 函数 — 候选产物
// ============================================================

export async function apiGetCandidates(
  workspaceId: string,
  runId: string,
): Promise<CandidateListResponse> {
  const res = await http.get<CandidateListResponse>(
    `/research/workspaces/${workspaceId}/runs/${runId}/candidates`,
  );
  return res.data;
}

export async function apiGetCandidateDetail(
  workspaceId: string,
  runId: string,
  candidateId: string,
): Promise<CandidateDetailResponse> {
  const res = await http.get<CandidateDetailResponse>(
    `/research/workspaces/${workspaceId}/runs/${runId}/candidates/${candidateId}`,
  );
  return res.data;
}

// ============================================================
// API 函数 — DerivedDataset
// ============================================================

export async function apiCreateDataset(
  workspaceId: string,
  body: { artifact_id: string; name: string; summary?: string; tags?: string[] },
): Promise<DerivedDataset> {
  const res = await http.post<DerivedDataset>(
    `/research/workspaces/${workspaceId}/derived-datasets`,
    body,
  );
  return res.data;
}

export async function apiListDatasets(
  workspaceId: string,
): Promise<DatasetListResponse> {
  const res = await http.get<DatasetListResponse>(
    `/research/workspaces/${workspaceId}/derived-datasets`,
  );
  return res.data;
}

export async function apiGetDataset(
  workspaceId: string,
  datasetId: string,
): Promise<DatasetDetail> {
  const res = await http.get<DatasetDetail>(
    `/research/workspaces/${workspaceId}/derived-datasets/${datasetId}`,
  );
  return res.data;
}

export async function apiUpdateDatasetMetadata(
  workspaceId: string,
  datasetId: string,
  body: { name?: string; summary?: string; tags?: string[] },
): Promise<DerivedDataset> {
  const res = await http.patch<DerivedDataset>(
    `/research/workspaces/${workspaceId}/derived-datasets/${datasetId}`,
    body,
  );
  return res.data;
}

export async function apiListDatasetVersions(
  workspaceId: string,
  datasetId: string,
): Promise<DatasetVersionListResponse> {
  const res = await http.get<DatasetVersionListResponse>(
    `/research/workspaces/${workspaceId}/derived-datasets/${datasetId}/versions`,
  );
  return res.data;
}

export async function apiGetDatasetVersion(
  workspaceId: string,
  datasetId: string,
  versionNumber: number,
): Promise<DatasetVersionDetail> {
  const res = await http.get<DatasetVersionDetail>(
    `/research/workspaces/${workspaceId}/derived-datasets/${datasetId}/versions/${versionNumber}`,
  );
  return res.data;
}

// ============================================================
// API 函数 — ResearchView
// ============================================================

export async function apiCreateView(
  workspaceId: string,
  body: { artifact_id: string; name: string; caption?: string; display_order?: number },
): Promise<View> {
  const res = await http.post<View>(
    `/research/workspaces/${workspaceId}/views`,
    body,
  );
  return res.data;
}

export async function apiListViews(
  workspaceId: string,
): Promise<ViewListResponse> {
  const res = await http.get<ViewListResponse>(
    `/research/workspaces/${workspaceId}/views`,
  );
  return res.data;
}

export async function apiGetView(
  workspaceId: string,
  viewId: string,
): Promise<ViewDetail> {
  const res = await http.get<ViewDetail>(
    `/research/workspaces/${workspaceId}/views/${viewId}`,
  );
  return res.data;
}

export async function apiUpdateViewMetadata(
  workspaceId: string,
  viewId: string,
  body: { name?: string; caption?: string; display_order?: number },
): Promise<View> {
  const res = await http.patch<View>(
    `/research/workspaces/${workspaceId}/views/${viewId}`,
    body,
  );
  return res.data;
}

export async function apiListViewVersions(
  workspaceId: string,
  viewId: string,
): Promise<ViewVersionListResponse> {
  const res = await http.get<ViewVersionListResponse>(
    `/research/workspaces/${workspaceId}/views/${viewId}/versions`,
  );
  return res.data;
}

export async function apiGetViewVersion(
  workspaceId: string,
  viewId: string,
  versionNumber: number,
): Promise<ViewVersionDetail> {
  const res = await http.get<ViewVersionDetail>(
    `/research/workspaces/${workspaceId}/views/${viewId}/versions/${versionNumber}`,
  );
  return res.data;
}

export function getViewImageUrl(
  workspaceId: string,
  viewId: string,
  versionNumber: number,
): string {
  const baseURL = (import.meta as any).env?.VITE_API_BASE_URL ?? '/api/v1';
  return `${baseURL}/research/workspaces/${workspaceId}/views/${viewId}/versions/${versionNumber}/image`;
}

// ============================================================
// API 函数 — Insight
// ============================================================

export async function apiListInsights(
  workspaceId: string,
): Promise<InsightListResponse> {
  const res = await http.get<InsightListResponse>(
    `/research/workspaces/${workspaceId}/insights`,
  );
  return res.data;
}

export async function apiGetInsight(
  workspaceId: string,
  insightId: string,
): Promise<InsightDetail> {
  const res = await http.get<InsightDetail>(
    `/research/workspaces/${workspaceId}/insights/${insightId}`,
  );
  return res.data;
}

export async function apiUpdateInsightMetadata(
  workspaceId: string,
  insightId: string,
  body: { name: string },
): Promise<Insight> {
  const res = await http.patch<Insight>(
    `/research/workspaces/${workspaceId}/insights/${insightId}`,
    body,
  );
  return res.data;
}

export async function apiListInsightVersions(
  workspaceId: string,
  insightId: string,
): Promise<InsightVersionListResponse> {
  const res = await http.get<InsightVersionListResponse>(
    `/research/workspaces/${workspaceId}/insights/${insightId}/versions`,
  );
  return res.data;
}

// ============================================================
// API 函数 — Insight Candidate
// ============================================================

export async function apiListInsightCandidates(
  workspaceId: string,
  runId: string,
): Promise<InsightCandidateListResponse> {
  const res = await http.get<InsightCandidateListResponse>(
    `/research/workspaces/${workspaceId}/runs/${runId}/insight-candidates`,
  );
  return res.data;
}

export async function apiAcceptCandidate(
  workspaceId: string,
  runId: string,
  candidateId: string,
): Promise<Insight> {
  const res = await http.post<Insight>(
    `/research/workspaces/${workspaceId}/runs/${runId}/insight-candidates/${candidateId}/accept`,
  );
  return res.data;
}

export async function apiModifyCandidate(
  workspaceId: string,
  runId: string,
  candidateId: string,
  body: {
    conclusion?: string;
    scope?: string;
    evidence_refs?: Array<Record<string, unknown>>;
    method_refs?: Array<Record<string, unknown>>;
    confidence_level?: string;
    limitations?: string;
    evidence_source_label?: string;
    modification_note: string;
  },
): Promise<Insight> {
  const res = await http.post<Insight>(
    `/research/workspaces/${workspaceId}/runs/${runId}/insight-candidates/${candidateId}/modify`,
    body,
  );
  return res.data;
}

export async function apiRejectCandidate(
  workspaceId: string,
  runId: string,
  candidateId: string,
  reason?: string,
): Promise<void> {
  await http.post(
    `/research/workspaces/${workspaceId}/runs/${runId}/insight-candidates/${candidateId}/reject`,
    { reason: reason ?? null },
  );
}

export async function apiRejectAnyCandidate(
  workspaceId: string,
  runId: string,
  candidateId: string,
  reason?: string,
): Promise<void> {
  await http.post(
    `/research/workspaces/${workspaceId}/runs/${runId}/candidates/${candidateId}/reject`,
    { reason: reason ?? null },
  );
}

export async function apiDeleteInsight(
  workspaceId: string,
  insightId: string,
): Promise<void> {
  await http.delete(`/research/workspaces/${workspaceId}/insights/${insightId}`);
}

export async function apiDeleteDataset(
  workspaceId: string,
  datasetId: string,
): Promise<void> {
  await http.delete(`/research/workspaces/${workspaceId}/derived-datasets/${datasetId}`);
}

export async function apiDeleteView(
  workspaceId: string,
  viewId: string,
): Promise<void> {
  await http.delete(`/research/workspaces/${workspaceId}/views/${viewId}`);
}

// ============================================================
// API 函数 — 产物列表
// ============================================================

export async function apiListProducts(
  workspaceId: string,
): Promise<ProductListResponse> {
  const res = await http.get<ProductListResponse>(
    `/research/workspaces/${workspaceId}/products`,
  );
  return res.data;
}

// ============================================================
// API 函数 — ResearchCatalog
// ============================================================

export async function apiSearchCatalog(
  query: string,
  workspaceId?: string,
): Promise<CatalogSearchResponse> {
  const params: Record<string, string> = { query };
  if (workspaceId) params.workspace_id = workspaceId;
  const res = await http.get<CatalogSearchResponse>(
    `/research/catalog/search`,
    { params },
  );
  return res.data;
}

File: apps/api/schemas/research_products.py
"""研究产物 API 的 Pydantic 请求/响应模型。

从 apps/api/routers/research_products.py 提取（P2-C5）。
"""

from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

# ---- DerivedDataset ----


class CreateDatasetRequest(BaseModel):
    """创建 DerivedDataset 请求。"""

    artifact_id: UUID
    name: str
    summary: str | None = None
    tags: list[str] = Field(default_factory=list)


class UpdateDatasetMetadataRequest(BaseModel):
    """编辑 DerivedDataset 元数据请求。"""

    name: str | None = None
    summary: str | None = None
    tags: list[str] | None = None


class DatasetResponse(BaseModel):
    dataset_id: UUID
    name: str
    status: str
    current_version: int
    workspace_id: UUID


# ---- ResearchView ----


class CreateViewRequest(BaseModel):
    """创建 ResearchView 请求。"""

    artifact_id: UUID
    name: str
    caption: str | None = None
    display_order: int = 0


class UpdateViewMetadataRequest(BaseModel):
    """编辑 ResearchView 元数据请求。"""

    name: str | None = None
    caption: str | None = None
    display_order: int | None = None


class ViewResponse(BaseModel):
    view_id: UUID
    name: str
    status: str
    current_version: int
    caption: str | None = None
    display_order: int = 0


# ---- Insight ----


class UpdateInsightMetadataRequest(BaseModel):
    """编辑 Insight 元数据请求。"""

    name: str


class InsightResponse(BaseModel):
    insight_id: UUID
    name: str
    status: str
    current_version: int


# ---- 候选产物 ----


class ModifyCandidateRequest(BaseModel):
    """修改 Insight 候选请求。"""

    conclusion: str | None = None
    scope: str | None = None
    evidence_refs: list[dict[str, Any]] | None = None
    method_refs: list[dict[str, Any]] | None = None
    confidence_level: str | None = None
    limitations: str | None = None
    evidence_source_label: str | None = None
    modification_note: str


class RejectCandidateRequest(BaseModel):
    """拒绝 Insight 候选请求。"""

    reason: str | None = None


# ---- 响应模型（list / detail） ----


class CandidateListItemResponse(BaseModel):
    """候选产物列表项。"""

    candidate_type: str
    source_artifact_id: str | None = None
    candidate_id: str
    source_run_id: str
    source_step_id: str | None = None
    step_name: str | None = None
    step_status: str | None = None
    preview_data: Any = None
    status: str | None = None
    error_reason: str | None = None


class CandidateListResponse(BaseModel):
    """候选产物列表。"""

    items: list[CandidateListItemResponse]


class CandidateDetailResponse(BaseModel):
    """候选产物详情。"""

    candidate_type: str
    candidate_id: str
    source_run_id: str
    source_step_id: str | None = None
    preview_data: Any = None


class DatasetListItemResponse(BaseModel):
    """DerivedDataset 列表项。"""

    dataset_id: str
    name: str
    status: str
    current_version: int
    workspace_id: str


class DatasetListResponse(BaseModel):
    """DerivedDataset 列表。"""

    items: list[DatasetListItemResponse]


class DatasetDetailResponse(BaseModel):
    """DerivedDataset 详情。"""

    dataset_id: str
    workspace_id: str
    name: str
    summary: str | None = None
    tags: list[str] = Field(default_factory=list)
    status: str
    current_version: int
    source_run_id: str
    source_snapshot_id: str | None = None
    current_version_data: Any = None


class DatasetVersionListItemResponse(BaseModel):
    """DerivedDataset 版本列表项。"""

    version_id: str
    dataset_id: str
    version_number: int
    content_hash: str | None = None
    created_at: str | None = None


class DatasetVersionListResponse(BaseModel):
    """DerivedDataset 版本列表。"""

    items: list[DatasetVersionListItemResponse]


class DatasetVersionDetailResponse(BaseModel):
    """DerivedDataset 版本详情。"""

    version_id: str
    dataset_id: str
    version_number: int
    metadata_content: Any = None
    points_content: Any = None
    series_content: Any = None
    field_manifest: Any = None
    content_hash: str | None = None
    source_run_id: str
    source_step_id: str | None = None
    source_artifact_id: str | None = None
    created_at: str | None = None


class ViewListItemResponse(BaseModel):
    """ResearchView 列表项。"""

    view_id: str
    name: str
    status: str
    current_version: int
    caption: str | None = None
    display_order: int = 0


class ViewListResponse(BaseModel):
    """ResearchView 列表。"""

    items: list[ViewListItemResponse]


class ViewDetailResponse(BaseModel):
    """ResearchView 详情。"""

    view_id: str
    workspace_id: str
    name: str
    caption: str | None = None
    display_order: int = 0
    status: str
    current_version: int
    source_run_id: str
    current_version_info: Any = None


class ViewVersionListItemResponse(BaseModel):
    """ResearchView 版本列表项。"""

    version_id: str
    view_id: str
    version_number: int
    image_storage_path: str | None = None
    image_format: str | None = None
    created_at: str | None = None


class ViewVersionListResponse(BaseModel):
    """ResearchView 版本列表。"""

    items: list[ViewVersionListItemResponse]


class ViewVersionDetailResponse(BaseModel):
    """ResearchView 版本详情。"""

    version_id: str
    view_id: str
    version_number: int
    image_storage_path: str | None = None
    image_format: str | None = None
    image_width: int | None = None
    image_height: int | None = None
    image_content_hash: str | None = None
    chart_code_artifact_id: str | None = None
    image_digest: str | None = None
    source_run_id: str
    source_step_id: str | None = None
    source_artifact_id: str | None = None
    bound_dataset_version_id: str | None = None
    chart_description: str | None = None
    created_at: str | None = None


class InsightListItemResponse(BaseModel):
    """Insight 列表项。"""

    insight_id: str
    name: str
    status: str
    current_version: int


class InsightListResponse(BaseModel):
    """Insight 列表。"""

    items: list[InsightListItemResponse]


class InsightDetailResponse(BaseModel):
    """Insight 详情。"""

    insight_id: str
    workspace_id: str
    name: str
    status: str
    current_version: int
    source_run_id: str | None = None
    current_version_data: Any = None


class InsightVersionListItemResponse(BaseModel):
    """Insight 版本列表项。"""

    version_id: str
    insight_id: str
    version_number: int
    is_modified: bool | None = None
    created_at: str | None = None


class InsightVersionListResponse(BaseModel):
    """Insight 版本列表。"""

    items: list[InsightVersionListItemResponse]


class InsightCandidateListItemResponse(BaseModel):
    """Insight 候选列表项。"""

    candidate_id: str
    run_id: str
    step_id: str | None = None
    status: str
    conclusion: str | None = None
    evidence_source_label: str | None = None
    created_at: str | None = None


class InsightCandidateListResponse(BaseModel):
    """Insight 候选列表。"""

    items: list[InsightCandidateListItemResponse]


class ProductListItemResponse(BaseModel):
    """产物列表项。"""

    product_type: str
    product_id: str
    name: str
    status: str
    current_version: int


class ProductListResponse(BaseModel):
    """产物列表。"""

    items: list[ProductListItemResponse]
